File: next/dweb-browser/MicroService/src/Ipc/Message/IpcRequest.cs
﻿namespace DwebBrowser.MicroService.Message;

public class IpcRequest : IpcMessage
{
    public int ReqId { get; init; }
    public override IPC_MESSAGE_TYPE Type { get; set; } = IPC_MESSAGE_TYPE.REQUEST;
    public IpcMethod Method { get; init; }
    public string Url { get; init; }
    public IpcHeaders Headers { get; set; }
    public IpcBody Body { get; set; }
    public Ipc Ipc { get; init; }

    public IpcRequest(int req_id, string url, IpcMethod method, IpcHeaders headers, IpcBody body, Ipc ipc)
    {
        ReqId = req_id;
        Url = url;
        Method = method;
        Headers = headers;
        Body = body;
        Ipc = ipc;

        Uri = new Uri(url);
    }

    public Uri Uri { get; init; }

    public static IpcRequest FromText(int req_id, string url, IpcMethod method, IpcHeaders headers, string text, Ipc ipc) =>
        new IpcRequest(req_id, url, method ?? IpcMethod.Get, headers ?? new IpcHeaders(), IpcBodySender.From(text, ipc), ipc);

    public static IpcRequest FromBinary(int req_id, string url, IpcMethod method, IpcHeaders headers, byte[] binary, Ipc ipc) =>
        new IpcRequest(
            req_id,
            url,
            method,
            (headers ??= new IpcHeaders()).Also(it =>
            {
                it.Init("Content-Type", "application/octet-stream");
                it.Init("Content-Length", binary.Length.ToString());
            }),
            IpcBodySender.From(binary, ipc),
            ipc);

    public static IpcRequest FromStream(
        int req_id,
        string url,
        IpcMethod method,
        IpcHeaders headers,
        Stream stream,
        Ipc ipc,
        long? size
        ) => new IpcRequest(
                req_id,
                url,
                method,
                (headers ??= new IpcHeaders()).Also(it =>
                {
                    it.Init("Content-Type", "application/octet-stream");

                    if (size != null)
                    {
                        headers.Init("Content-Length", size.ToString()!);
                    }
                }),
                IpcBodySender.From(stream, ipc),
                ipc);

    //public static IpcRequest FromRequest(int req_id, HttpRequestMessage request, Ipc ipc) =>
    //    new IpcRequest(
    //        req_id,
    //        request.RequestUri!.ToString(),
    //        IpcMethod.From(request.Method),
    //        new IpcHeaders(request.Headers),
    //        (request.Method.Method is "GET" or "HEAD")
    //            ? IpcBodySender.From("", ipc)
    //            :
    //        request.Content switch
    //        {
    //            null => IpcBodySender.From("", ipc),
    //            _ => IpcBodySender.From(request.Content.ReadAsStream(), ipc)
    //        },
    //        ipc
    //        );
    public static IpcRequest FromRequest(int req_id, HttpRequestMessage request, Ipc ipc)
    {

        //switch (self.Content.ToString())
        //{
        //    case "System.Net.Http.StringContent":
        //        var result = JsonSerializer.Deserialize<T>(await self.TextAsync())!;
        //        Console.WriteLine($"result: {result}");
        //        return result;
        //    case "System.IO.MemoryStream":
        //    case "System.Net.Http.StreamContent":
        //        return JsonSerializer.Deserialize<T>(await self.StreamAsync())!;
        //}
        var body = request.Content.ToString() switch
        {
            "System.Net.Http.EmptyReadStream" => IpcBodySender.From("", ipc),
            "System.Net.Http.StringContent" => IpcBodySender.From(request.Content.ReadAsStringAsync().Result, ipc),
            "System.Net.Http.ByteArrayContent" => IpcBodySender.From(request.Content.ReadAsByteArrayAsync().Result, ipc),
            _ => IpcBodySender.From(request.Content.ReadAsStream(), ipc)
        };
        Console.WriteLine(request.RequestUri?.ToString());
        try
        {
            var ipcRequest = new IpcRequest(
            req_id,
            request.RequestUri?.ToString() ?? "",
            IpcMethod.From(request.Method),
            new IpcHeaders(request.Headers),
            (request.Method.Method is "GET" or "HEAD")
                ? IpcBodySender.From("", ipc)
                : body,
            ipc
            );
            return ipcRequest;
        }
        catch(Exception e)
        {
            Console.WriteLine($"e {e.StackTrace}");
            Console.WriteLine($"e {e.Message}");
            throw e;
        }
    }


    public HttpRequestMessage ToRequest() =>
        new HttpRequestMessage(new HttpMethod(Method.method), new Uri(Url)).Also(it =>
            {
                switch (Body.Raw)
                {
                    case string body:
                        it.Content = new StringContent(body);
                        break;
                    case byte[] body:
                        it.Content = new ByteArrayContent(body);
                        break;
                    case Stream body:
                        it.Content = new StreamContent(body);
                        break;
                    default:
                        throw new Exception($"invalid body to request: {Body.Raw}");
                }

                foreach (var entry in Headers.GetEnumerator())
                {
                    if (entry.Key.StartsWith("Content", true, null))
                    {
                        it.Content.Headers.Add(entry.Key, entry.Value);
                    }
                    else
                    {
                        it.Headers.TryAddWithoutValidation(entry.Key, entry.Value);
                    }

                }
            });

    public IpcReqMessage LazyIpcReqMessage
    {
        get
        {
            return new Lazy<IpcReqMessage>(new Func<IpcReqMessage>(() =>
                new IpcReqMessage(
                    ReqId,
                    Method,
                    Url,
                    Headers.GetEnumerator().ToDictionary(k => k.Key, v => v.Value),
                    Body.MetaBody)), true).Value;
        }
    }

    public override string ToString() => $"#IpcRequest/{Method.method}/{Url}";
}

[JsonConverter(typeof(IpcReqMessageConverter))]
public class IpcReqMessage : IpcMessage
{
    public override IPC_MESSAGE_TYPE Type { get; set; } = IPC_MESSAGE_TYPE.REQUEST;

    public int ReqId { get; set; }
    public IpcMethod Method { get; set; }
    public string Url { get; set; }
    public Dictionary<string, string> Headers { get; set; }
    public SMetaBody MetaBody { get; set; }

    public IpcReqMessage(
        int req_id,
        IpcMethod method,
        string url,
        Dictionary<String, String> headers,
        SMetaBody metaBody)
    {
        ReqId = req_id;
        Method = method;
        Url = url;
        Headers = headers;
        MetaBody = metaBody;
    }

    /// <summary>
    /// Serialize IpcReqMessage
    /// </summary>
    /// <returns>JSON string representation of the IpcReqMessage</returns>
    public override string ToJson() => JsonSerializer.Serialize(this, new JsonSerializerOptions { IncludeFields = true });

    /// <summary>
    /// Deserialize IpcReqMessage
    /// </summary>
    /// <param name="json">JSON string representation of IpcReqMessage</param>
    /// <returns>An instance of a IpcReqMessage object.</returns>
    public static IpcReqMessage? FromJson(string json) => JsonSerializer.Deserialize<IpcReqMessage>(json, new JsonSerializerOptions { IncludeFields = true });
}

#region IpcReqMessage序列化反序列化
sealed class IpcReqMessageConverter : JsonConverter<IpcReqMessage>
{
    public override bool CanConvert(Type typeToConvert) =>
        typeToConvert.GetMethod("ToJson") != null && typeToConvert.GetMethod("FromJson") != null;


    public override IpcReqMessage? Read(
        ref Utf8JsonReader reader,
        Type typeToConvert,
        JsonSerializerOptions options)
    {
        if (reader.TokenType != JsonTokenType.StartObject)
            throw new JsonException("Expected StartObject token");

        int req_id = default;
        IPC_MESSAGE_TYPE type = default;
        string url = default;
        IpcMethod method = default;
        SMetaBody metaBody = default;
        var headers = new Dictionary<string, string>();
        while (reader.Read())
        {
            if (reader.TokenType == JsonTokenType.EndObject)
                return new IpcReqMessage(req_id, method ?? IpcMethod.Get, url ?? "", headers, metaBody);

            if (reader.TokenType != JsonTokenType.PropertyName)
                throw new JsonException("Expected PropertyName token");

            var propName = reader.GetString();

            reader.Read();

            switch (propName)
            {
                case "req_id":
                    req_id = reader.GetInt32();
                    break;
                case "type":
                    type = (IPC_MESSAGE_TYPE)reader.GetInt16();
                    break;
                case "url":
                    url = reader.GetString() ?? "";
                    break;
                case "method":
                    method = new IpcMethod(reader.GetString() ?? "GET");
                    break;
                case "metaBody":
                    SMetaBody.IPC_META_BODY_TYPE mtype = default;
                    int senderUid = default;
                    string data = default;
                    string? stream_id = null;
                    int? receiverUid = null;
                    string metaId = default;

                    while (reader.Read())
                    {
                        if (reader.TokenType == JsonTokenType.StartObject)
                        {
                            continue;
                        }

                        if (reader.TokenType == JsonTokenType.EndObject)
                        {
                            metaBody = new SMetaBody(mtype, senderUid, data ?? "", stream_id, receiverUid) { MetaId = metaId ?? "" };
                            break;
                        }


                        if (reader.TokenType != JsonTokenType.PropertyName)
                            throw new JsonException("Expected PropertyName token");

                        var mpropName = reader.GetString();

                        reader.Read();

                        switch (mpropName)
                        {
                            case "type":
                                mtype = (SMetaBody.IPC_META_BODY_TYPE)reader.GetInt64();
                                break;
                            case "senderUid":
                                senderUid = reader.GetInt32();
                                break;
                            case "data":
                                data = reader.GetString() ?? "";
                                break;
                            case "streamId":
                                stream_id = reader.GetString() ?? null;
                                break;
                            case "receiverUid":
                                receiverUid = reader.GetInt32();
                                break;
                            case "metaId":
                                metaId = reader.GetString() ?? "";
                                break;
                        }
                    }

                    break;
                case "headers":
                    while (reader.Read())
                    {
                        if (reader.TokenType == JsonTokenType.StartObject)
                        {
                            continue;
                        }

                        if (reader.TokenType == JsonTokenType.EndObject)
                        {
                            break;
                        }

                        var memberName = reader.GetString();

                        reader.Read();

                        if (memberName != null)
                        {
                            headers.Add(memberName, reader.GetString() ?? "");
                        }
                    }

                    break;
            }
        }

        throw new JsonException("Expected EndObject token");
    }

    public override void Write(
        Utf8JsonWriter writer,
        IpcReqMessage value,
        JsonSerializerOptions options)
    {
        writer.WriteStartObject();

        writer.WriteNumber("req_id", value.ReqId);
        writer.WriteNumber("type", (int)value.Type);
        writer.WriteString("method", value.Method.method);
        writer.WriteString("url", value.Url);
        writer.WriteString("metaBody", value.MetaBody.ToJson());

        // dictionary
        writer.WritePropertyName("headers");
        writer.WriteStartObject();

        foreach ((string key, string keyValue) in value.Headers)
        {

            writer.WriteString(key, keyValue);
        }

        writer.WriteEndObject();

        writer.WriteEndObject();
    }
}
#endregion

File: desktop-dev/src/core/ipc/IpcRequest.ts
import { once } from "../../helper/$once.ts";
import { $Binary, binaryToU8a, isBinary } from "../../helper/binaryHelper.ts";
import { httpMethodCanOwnBody } from "../../helper/httpHelper.ts";
import { parseUrl } from "../../helper/urlHelper.ts";
import type { $BodyData, IpcBody } from "./IpcBody.ts";
import { IpcBodySender } from "./IpcBodySender.ts";
import { IpcHeaders } from "./IpcHeaders.ts";
import type { MetaBody } from "./MetaBody.ts";
import {
  IPC_MESSAGE_TYPE,
  IPC_METHOD,
  IpcMessage,
  toIpcMethod,
} from "./const.ts";
import type { Ipc } from "./ipc.ts";

export class IpcRequest extends IpcMessage<IPC_MESSAGE_TYPE.REQUEST> {
  constructor(
    readonly req_id: number,
    readonly url: string,
    readonly method: IPC_METHOD,
    readonly headers: IpcHeaders,
    readonly body: IpcBody,
    readonly ipc: Ipc
  ) {
    super(IPC_MESSAGE_TYPE.REQUEST);
    if (body instanceof IpcBodySender) {
      IpcBodySender.$usableByIpc(ipc, body);
    }
  }

  #parsed_url?: URL;
  get parsed_url() {
    return (this.#parsed_url ??= parseUrl(this.url));
  }

  static fromText(
    req_id: number,
    url: string,
    method: IPC_METHOD = IPC_METHOD.GET,
    headers = new IpcHeaders(),
    text: string,
    ipc: Ipc
  ) {
    // 这里 content-length 默认不写，因为这是要算二进制的长度，我们这里只有在字符串的长度，不是一个东西
    return new IpcRequest(
      req_id,
      url,
      method,
      headers,
      IpcBodySender.fromText(text, ipc),
      ipc
    );
  }
  static fromBinary(
    req_id: number,
    url: string,
    method: IPC_METHOD = IPC_METHOD.GET,
    headers = new IpcHeaders(),
    binary: $Binary,
    ipc: Ipc
  ) {
    headers.init("Content-Type", "application/octet-stream");
    headers.init("Content-Length", binary.byteLength + "");

    return new IpcRequest(
      req_id,
      url,
      method,
      headers,
      IpcBodySender.fromBinary(binaryToU8a(binary), ipc),
      ipc
    );
  }
  // 如果需要发送stream数据 一定要使用这个方法才可以传递数据否则数据无法传递
  static fromStream(
    req_id: number,
    url: string,
    method: IPC_METHOD = IPC_METHOD.GET,
    headers = new IpcHeaders(),
    stream: ReadableStream<Uint8Array>,
    ipc: Ipc
  ) {
    headers.init("Content-Type", "application/octet-stream");

    return new IpcRequest(
      req_id,
      url,
      method,
      headers,
      IpcBodySender.fromStream(stream, ipc),
      ipc
    );
  }

  static fromRequest(
    req_id: number,
    ipc: Ipc,
    url: string,
    init: {
      method?: string;
      body?: /* json+text */
      | null
        | string
        /* base64 */
        | Uint8Array
        /* stream+base64 */
        | ReadableStream<Uint8Array>;
      headers?: IpcHeaders | HeadersInit;
    } = {}
  ) {
    const method = toIpcMethod(init.method);
    const headers =
      init.headers instanceof IpcHeaders
        ? init.headers
        : new IpcHeaders(init.headers);

    let ipcBody: IpcBody;
    if (isBinary(init.body)) {
      ipcBody = IpcBodySender.fromBinary(init.body, ipc);
    } else if (init.body instanceof ReadableStream) {
      ipcBody = IpcBodySender.fromStream(init.body, ipc);
    } else {
      ipcBody = IpcBodySender.fromText(init.body ?? "", ipc);
    }

    return new IpcRequest(req_id, url, method, headers, ipcBody, ipc);
  }

  /**
   * 判断是否是双工协议
   *
   * 比如目前双工协议可以由 WebSocket 来提供支持
   */
  isDuplex() {
    return (
      this.method === "GET" &&
      this.headers.get("Upgrade")?.toLowerCase() === "websocket"
    );
  }

  toRequest() {
    let { method } = this;
    let body: undefined | $BodyData;
    const isWebSocket = this.isDuplex();
    if (isWebSocket) {
      method = IPC_METHOD.POST; // new Request 如果要携带body，method 不可以是 GET
      body = this.body.raw;
    } else if (httpMethodCanOwnBody(method)) {
      body = this.body.raw;
    }
    const init = {
      method,
      headers: this.headers,
      body,
      duplex: body instanceof ReadableStream ? "half" : undefined,
    };
    if (body) {
      Reflect.set(init, "duplex", "half");
    }
    const request = new Request(this.url, init);
    if (isWebSocket) {
      Object.defineProperty(request, "method", {
        configurable: true,
        enumerable: true,
        writable: false,
        value: "GET",
      });
    }

    return request;
  }

  readonly ipcReqMessage = once(
    () =>
      new IpcReqMessage(
        this.req_id,
        this.method,
        this.url,
        this.headers.toJSON(),
        this.body.metaBody
      )
  );

  toJSON() {
    const { method } = this;
    // let body: undefined | $BodyData;
    if ((method === IPC_METHOD.GET || method === IPC_METHOD.HEAD) === false) {
      // body = this.body.raw;
      return new IpcReqMessage(
        this.req_id,
        this.method,
        this.url,
        this.headers.toJSON(),
        this.body.metaBody
      );
    }
    return this.ipcReqMessage();
  }
}

export class IpcReqMessage extends IpcMessage<IPC_MESSAGE_TYPE.REQUEST> {
  constructor(
    readonly req_id: number,
    readonly method: IPC_METHOD,
    readonly url: string,
    readonly headers: Record<string, string>,
    readonly metaBody: MetaBody
  ) {
    super(IPC_MESSAGE_TYPE.REQUEST);
  }
}
